### PR TITLE
[cmake] Export cmake configure files

### PR DESCRIPTION
**Pull request by JackBoosY**
So far, the use of wxwidgets in cmake is still through `FindWxwidgets.cmake` provided by cmake.
However, this cmake file has existed for a long time and the code changes will not be synchronized immediately.

I recommend exporting all targets to make it easier and safer to use wxwidgets.

There are still many inappropriate aspects of this PR, and I hope that the official will consider this method.

Related: https://github.com/microsoft/vcpkg/pull/17111.

### CODE REVIEW

**Comment by vadz** [comment]
Sorry @MaartenBent but this one is also for you I'm afraid (and/or maybe @TcT2k?).

**Comment by MaartenBent** [comment]
Providing a `wxwidgets-config.cmake` file seems indeed like a good idea. I'm not familiar with config files, but I'll have a closer look at this PR later.

**Comment by alandefreitas** [comment]
That's amazing. Just to emphasize how useful this is, a simple `wxwidgets-config.cmake` exporting all install targets, which already exist, could solve many small problems with `FindWxwidgets.cmake` + `UsewxWidgets.cmake`:

- `wxwidgets-config.cmake` is going to provide Modern CMake / targets always synchronized with the current wxwidgets version
- `FindWxwidgets.cmake` [looks](https://github.com/Kitware/CMake/blob/ed7b66561e9e2e748bac0b06fd3cb8782e4fc75e/Modules/FindwxWidgets.cmake#L780) for variants of `wx-config(-<version>)?` and `wxgtk3u-<version>-config` but the default CMake installer doesn't always installs these files (only `osx_cocoa-unicode-static-3.1` here) so nothing is found (without a patch)
- The flags from `wx-config` [are split](https://github.com/Kitware/CMake/blob/ed7b66561e9e2e748bac0b06fd3cb8782e4fc75e/Modules/FindwxWidgets.cmake#L812) with `separate_arguments(wxWidgets_CXX_FLAGS_LIST NATIVE_COMMAND "${wxWidgets_CXX_FLAGS}")`, which doesn't work if there's any whitespace in the install directory

**Comment by alandefreitas** [comment]
This is a very minor detail, but it seems like CMake [prefers](https://cmake.org/cmake/help/latest/command/find_package.html) `wxWidgetsConfig.cmake` over `wxWidgets-config.cmake`, and the `-config.cmake` version expects a lower case package name. 

In many cases, everything is treated as case-insensitive, although lower case usually has a preference, so it's worth testing. Most package managers will probably need to lowercase the package name also.

**Comment by MaartenBent** [comment]
I had a look at the implementation, it needed some small changes to be better compatible with the build system. I included a patch below, or look at the commits in [this](https://github.com/MaartenBent/wxWidgets/commits/cmake-export) branch. (note: I rebased this PR first). Maybe you can add them to this PR?

I have not tested if it actually works, only that the export files are created and installed.

A possible problem I noticed is that `wxWidgets-config.cmake.in` contains `unofficial-${TARGET_FILE}Targets.cmake` but the export files are not named like this. So I suspect this needs to be fixed.
Also do the export files have to be in `installdir/lib/platform/cmake` or can/should it be directly in `installdir/cmake`?

The changes in the patch:
- Using `wxPLATFORM_LIB_DIR` caused some problems on Windows (see AppVeyor logs) so I removed the generator expression from it.
- There is no need to install each 3th party built-in library separately (regex, scintilla, but also jpeg, tiff, png). I moved this code to `wx_set_builtin_target_properties`. We also have a `wx_install` macro that will disable the install targets if the user disables `wxBUILD_INSTALL`.
- Try too keep CMake files inside `build/cmake`, so I moved `wxWidgets-config.cmake.in`. Maybe have a look at renaming it, as @alandefreitas suggests.
- I simplified `wx_lib_include_directories`, `INCLUDE_POS` is not needed anymore.

<details><summary>Click to see patch</summary>
<p>

```diff
From 637c27e31b101d23a24a19dac88e33bbaf8172da Mon Sep 17 00:00:00 2001
From: Maarten Bent <MaartenBent@users.noreply.github.com>
Date: Sun, 11 Jul 2021 14:05:46 +0200
Subject: [PATCH 1/4] CMake: Remove generator expression from
 wxPLATFORM_LIB_DIR

Use it directly in set_target_properties.
---
 CMakeLists.txt              |  2 +-
 build/cmake/functions.cmake | 16 ++++++++--------
 build/cmake/init.cmake      |  5 ++---
 build/cmake/install.cmake   |  2 +-
 4 files changed, 12 insertions(+), 13 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 23b06ac1cf..52a74ef49f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -97,4 +97,4 @@ if(wxBUILD_SAMPLES AND wxUSE_GUI)
     set_directory_properties(PROPERTIES VS_STARTUP_PROJECT minimal)
 endif()
 configure_file(${CMAKE_CURRENT_LIST_DIR}/wxWidgets-config.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/wxWidgets-config.cmake @ONLY)
-install(FILES ${CMAKE_CURRENT_BINARY_DIR}/wxWidgets-config.cmake DESTINATION lib${wxPLATFORM_LIB_DIR}/cmake)
\ No newline at end of file
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/wxWidgets-config.cmake DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
\ No newline at end of file
diff --git a/build/cmake/functions.cmake b/build/cmake/functions.cmake
index 87a22f1c12..bb5871e1bd 100644
--- a/build/cmake/functions.cmake
+++ b/build/cmake/functions.cmake
@@ -97,9 +97,9 @@ function(wx_set_common_target_properties target_name)
     cmake_parse_arguments(wxCOMMON_TARGET_PROPS "DEFAULT_WARNINGS" "" "" ${ARGN})
 
     set_target_properties(${target_name} PROPERTIES
-        LIBRARY_OUTPUT_DIRECTORY "${wxOUTPUT_DIR}${wxPLATFORM_LIB_DIR}"
-        ARCHIVE_OUTPUT_DIRECTORY "${wxOUTPUT_DIR}${wxPLATFORM_LIB_DIR}"
-        RUNTIME_OUTPUT_DIRECTORY "${wxOUTPUT_DIR}${wxPLATFORM_LIB_DIR}"
+        LIBRARY_OUTPUT_DIRECTORY "${wxOUTPUT_DIR}$<1:/>${wxPLATFORM_LIB_DIR}"
+        ARCHIVE_OUTPUT_DIRECTORY "${wxOUTPUT_DIR}$<1:/>${wxPLATFORM_LIB_DIR}"
+        RUNTIME_OUTPUT_DIRECTORY "${wxOUTPUT_DIR}$<1:/>${wxPLATFORM_LIB_DIR}"
         )
 
     if(wxBUILD_PIC)
@@ -418,13 +418,13 @@ macro(wx_add_library name)
             set(runtime_dir "bin")
         endif()
         wx_install(TARGETS ${name} EXPORT ${name}Targets
-            LIBRARY DESTINATION "lib${wxPLATFORM_LIB_DIR}"
-            ARCHIVE DESTINATION "lib${wxPLATFORM_LIB_DIR}"
-            RUNTIME DESTINATION "${runtime_dir}${wxPLATFORM_LIB_DIR}"
+            LIBRARY DESTINATION "lib$<1:/>${wxPLATFORM_LIB_DIR}"
+            ARCHIVE DESTINATION "lib$<1:/>${wxPLATFORM_LIB_DIR}"
+            RUNTIME DESTINATION "${runtime_dir}$<1:/>${wxPLATFORM_LIB_DIR}"
             BUNDLE DESTINATION Applications/wxWidgets
             )
 
-        install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION lib${wxPLATFORM_LIB_DIR}/cmake)
+        install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
     endif()
 endmacro()
 
@@ -586,7 +586,7 @@ function(wx_set_builtin_target_properties target_name)
 
     wx_set_common_target_properties(${target_name} DEFAULT_WARNINGS)
     if(NOT wxBUILD_SHARED)
-        wx_install(TARGETS ${name} ARCHIVE DESTINATION "lib${wxPLATFORM_LIB_DIR}")
+        wx_install(TARGETS ${name} ARCHIVE DESTINATION "lib$<1:/>${wxPLATFORM_LIB_DIR}")
     endif()
 endfunction()
 
diff --git a/build/cmake/init.cmake b/build/cmake/init.cmake
index e1426afd42..4802b20054 100644
--- a/build/cmake/init.cmake
+++ b/build/cmake/init.cmake
@@ -153,8 +153,7 @@ if(WIN32_MSVC_NAMING)
         set(lib_suffix "_lib")
     endif()
 
-    # Include generator expression to suppress default Debug/Release pair
-    set(wxPLATFORM_LIB_DIR "$<1:/>${wxCOMPILER_PREFIX}${wxARCH_SUFFIX}${lib_suffix}")
+    set(wxPLATFORM_LIB_DIR "${wxCOMPILER_PREFIX}${wxARCH_SUFFIX}${lib_suffix}")
 endif()
 
 if(wxBUILD_CUSTOM_SETUP_HEADER_PATH)
@@ -171,7 +170,7 @@ else()
             set(lib_unicode)
         endif()
         set(wxSETUP_HEADER_PATH
-            ${wxOUTPUT_DIR}/${wxCOMPILER_PREFIX}${wxARCH_SUFFIX}${lib_suffix}/${wxBUILD_TOOLKIT}${lib_unicode})
+            ${wxOUTPUT_DIR}/${wxPLATFORM_LIB_DIR}/${wxBUILD_TOOLKIT}${lib_unicode})
         file(MAKE_DIRECTORY ${wxSETUP_HEADER_PATH}/wx)
         file(MAKE_DIRECTORY ${wxSETUP_HEADER_PATH}d/wx)
         set(wxSETUP_HEADER_FILE_DEBUG ${wxSETUP_HEADER_PATH}d/wx/setup.h)
diff --git a/build/cmake/install.cmake b/build/cmake/install.cmake
index ef971c7285..7902646ef4 100644
--- a/build/cmake/install.cmake
+++ b/build/cmake/install.cmake
@@ -32,7 +32,7 @@ endif()
 if(WIN32_MSVC_NAMING)
     wx_install(
         DIRECTORY "${wxSETUP_HEADER_PATH}"
-        DESTINATION "lib${wxPLATFORM_LIB_DIR}")
+        DESTINATION "lib/${wxPLATFORM_LIB_DIR}")
 else()
     wx_install(
         DIRECTORY "${wxSETUP_HEADER_PATH}"
-- 
2.32.0.windows.2


From f54d78c071237096aea0f8c6174c6f19ada04b9a Mon Sep 17 00:00:00 2001
From: Maarten Bent <MaartenBent@users.noreply.github.com>
Date: Sun, 11 Jul 2021 14:09:33 +0200
Subject: [PATCH 2/4] CMake: Use common function to install 3th party exports

Use wx_install instead of install.
Remove unneeded wx_install in wx_add_thirdparty_library.
---
 build/cmake/functions.cmake        | 16 ++++++----------
 build/cmake/lib/regex.cmake        |  9 ---------
 build/cmake/lib/stc/CMakeLists.txt | 10 ----------
 3 files changed, 6 insertions(+), 29 deletions(-)

diff --git a/build/cmake/functions.cmake b/build/cmake/functions.cmake
index bb5871e1bd..ceab750439 100644
--- a/build/cmake/functions.cmake
+++ b/build/cmake/functions.cmake
@@ -424,7 +424,7 @@ macro(wx_add_library name)
             BUNDLE DESTINATION Applications/wxWidgets
             )
 
-        install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
+        wx_install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
     endif()
 endmacro()
 
@@ -586,7 +586,11 @@ function(wx_set_builtin_target_properties target_name)
 
     wx_set_common_target_properties(${target_name} DEFAULT_WARNINGS)
     if(NOT wxBUILD_SHARED)
-        wx_install(TARGETS ${name} ARCHIVE DESTINATION "lib$<1:/>${wxPLATFORM_LIB_DIR}")
+        # Only install if we build as static libraries
+        wx_install(TARGETS ${name} EXPORT ${name}Targets
+            ARCHIVE DESTINATION "lib$<1:/>${wxPLATFORM_LIB_DIR}"
+            )
+        wx_install(EXPORT ${name}Targets NAMESPACE wx:: DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
     endif()
 endfunction()
 
@@ -642,14 +646,6 @@ function(wx_add_thirdparty_library var_name lib_name help_str)
         endif()
     endif()
 
-    if(${var_name} STREQUAL "builtin" AND NOT wxBUILD_SHARED)
-        # Only install if we build as static libraries
-        wx_install(TARGETS ${target_name}
-            LIBRARY DESTINATION lib
-            ARCHIVE DESTINATION lib
-            )
-    endif()
-
     set(wxTHIRD_PARTY_LIBRARIES ${wxTHIRD_PARTY_LIBRARIES} ${var_name} "${help_str}" PARENT_SCOPE)
 endfunction()
 
diff --git a/build/cmake/lib/regex.cmake b/build/cmake/lib/regex.cmake
index 7130287186..b28fc2c7ce 100644
--- a/build/cmake/lib/regex.cmake
+++ b/build/cmake/lib/regex.cmake
@@ -18,13 +18,4 @@ if(wxUSE_REGEX)
     target_include_directories(wxregex PRIVATE ${wxSETUP_HEADER_PATH} ${wxSOURCE_DIR}/include)
     set(REGEX_LIBRARIES wxregex)
     set(REGEX_INCLUDE_DIRS ${wxSOURCE_DIR}/src/regex)
-    
-    install(TARGETS wxregex EXPORT wxregexTargets
-        LIBRARY DESTINATION "lib${wxPLATFORM_LIB_DIR}"
-        ARCHIVE DESTINATION "lib${wxPLATFORM_LIB_DIR}"
-        RUNTIME DESTINATION "bin${wxPLATFORM_LIB_DIR}"
-        BUNDLE DESTINATION Applications/wxWidgets
-        )
-
-    install(EXPORT wxregexTargets NAMESPACE wx:: DESTINATION share/lib${wxPLATFORM_LIB_DIR}/cmake)
 endif()
diff --git a/build/cmake/lib/stc/CMakeLists.txt b/build/cmake/lib/stc/CMakeLists.txt
index 7cb592f39f..bd6acd858c 100644
--- a/build/cmake/lib/stc/CMakeLists.txt
+++ b/build/cmake/lib/stc/CMakeLists.txt
@@ -182,16 +182,6 @@ if(wxBUILD_PRECOMP)
     wx_target_enable_precomp(wxscintilla ${wxSCINTILLA_PREC_HEADER})
 endif()
 
-
-install(TARGETS wxscintilla EXPORT wxscintillaTargets
-    LIBRARY DESTINATION "lib${wxPLATFORM_LIB_DIR}"
-    ARCHIVE DESTINATION "lib${wxPLATFORM_LIB_DIR}"
-    RUNTIME DESTINATION "bin${wxPLATFORM_LIB_DIR}"
-    BUNDLE DESTINATION Applications/wxWidgets
-    )
-
-install(EXPORT wxscintillaTargets NAMESPACE wx:: DESTINATION lib${wxPLATFORM_LIB_DIR}/cmake)
-
 wx_add_library(wxstc ${STC_FILES})
 wx_lib_include_directories(wxstc PRIVATE
     ${wxSOURCE_DIR}/src/stc/scintilla/include
-- 
2.32.0.windows.2


From 783dfff30763a46f9b7c9e93fd84835a0128b009 Mon Sep 17 00:00:00 2001
From: Maarten Bent <MaartenBent@users.noreply.github.com>
Date: Sun, 11 Jul 2021 14:10:34 +0200
Subject: [PATCH 3/4] CMake: Move wxWidgets-config

Don't configure and install it from main script, but from install script.
---
 CMakeLists.txt                                                | 2 --
 build/cmake/install.cmake                                     | 4 ++++
 .../cmake/wxWidgets-config.cmake.in                           | 2 +-
 3 files changed, 5 insertions(+), 3 deletions(-)
 rename wxWidgets-config.cmake.in => build/cmake/wxWidgets-config.cmake.in (98%)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 52a74ef49f..d5147f1845 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -96,5 +96,3 @@ include(build/cmake/main.cmake)
 if(wxBUILD_SAMPLES AND wxUSE_GUI)
     set_directory_properties(PROPERTIES VS_STARTUP_PROJECT minimal)
 endif()
-configure_file(${CMAKE_CURRENT_LIST_DIR}/wxWidgets-config.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/wxWidgets-config.cmake @ONLY)
-install(FILES ${CMAKE_CURRENT_BINARY_DIR}/wxWidgets-config.cmake DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
\ No newline at end of file
diff --git a/build/cmake/install.cmake b/build/cmake/install.cmake
index 7902646ef4..72f89a0fda 100644
--- a/build/cmake/install.cmake
+++ b/build/cmake/install.cmake
@@ -55,6 +55,10 @@ else()
     )
 endif()
 
+configure_file(build/cmake/wxWidgets-config.cmake.in ${wxOUTPUT_DIR}/wxWidgets-config.cmake @ONLY)
+wx_install(FILES ${wxOUTPUT_DIR}/wxWidgets-config.cmake DESTINATION lib/${wxPLATFORM_LIB_DIR}/cmake)
+
+
 # uninstall target
 if(MSVC_IDE)
     set(UNINST_NAME UNINSTALL)
diff --git a/wxWidgets-config.cmake.in b/build/cmake/wxWidgets-config.cmake.in
similarity index 98%
rename from wxWidgets-config.cmake.in
rename to build/cmake/wxWidgets-config.cmake.in
index d1d92ba0a7..9dec819ffd 100644
--- a/wxWidgets-config.cmake.in
+++ b/build/cmake/wxWidgets-config.cmake.in
@@ -34,4 +34,4 @@ endif()
 list(APPEND TARGET_FILES wxregex wxscintilla wxbase wxcore wxgl wxxml wxhtml wxxrc wxstc wxwebview wxnet wxpropgrid wxqa wxribbon wxrichtext wxmedia wxadv wxaui)
 foreach (TARGET_FILE ${TARGET_FILES})
     include("${CMAKE_CURRENT_LIST_DIR}/unofficial-${TARGET_FILE}Targets.cmake")
-endforeach()
\ No newline at end of file
+endforeach()
-- 
2.32.0.windows.2


From 15a63aae35df6c66c3b2a540c6da744b0197fbde Mon Sep 17 00:00:00 2001
From: Maarten Bent <MaartenBent@users.noreply.github.com>
Date: Sun, 11 Jul 2021 14:10:43 +0200
Subject: [PATCH 4/4] CMake: Remove unneeded INCLUDE_POS and fix indenting

---
 build/cmake/functions.cmake | 12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/build/cmake/functions.cmake b/build/cmake/functions.cmake
index ceab750439..48c41bed50 100644
--- a/build/cmake/functions.cmake
+++ b/build/cmake/functions.cmake
@@ -497,17 +497,13 @@ macro(wx_lib_include_directories name)
         set(wxMONO_INCLUDE_DIRS_PUBLIC ${wxMONO_INCLUDE_DIRS_PUBLIC} PARENT_SCOPE)
         set(wxMONO_INCLUDE_DIRS_PRIVATE ${wxMONO_INCLUDE_DIRS_PRIVATE} PARENT_SCOPE)
     else()
-        set(INCLUDE_POS)
-        if (_LIB_INCLUDE_DIRS_PRIVATE)
-            set(INCLUDE_POS BEFORE)
-        endif()
         if (_LIB_INCLUDE_DIRS_PUBLIC)
-        foreach (PUBLIC_DIR ${_LIB_INCLUDE_DIRS_PUBLIC})
-            target_include_directories(${name};${INCLUDE_POS};PUBLIC;$<BUILD_INTERFACE:${PUBLIC_DIR}>)
-        endforeach()
+            foreach (PUBLIC_DIR ${_LIB_INCLUDE_DIRS_PUBLIC})
+                target_include_directories(${name};PUBLIC;$<BUILD_INTERFACE:${PUBLIC_DIR}>)
+            endforeach()
         endif()
         if (_LIB_INCLUDE_DIRS_PRIVATE)
-            target_include_directories(${name};${INCLUDE_POS};PRIVATE;${_LIB_INCLUDE_DIRS_PRIVATE})
+            target_include_directories(${name};BEFORE;PRIVATE;${_LIB_INCLUDE_DIRS_PRIVATE})
         endif()
         target_include_directories(${name};PUBLIC;$<INSTALL_INTERFACE:include>)
     endif()
-- 
2.32.0.windows.2

```
</p>
</details>

File: wxWidgets-config.cmake.in [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+if (@wxUSE_LIBPNG@ STREQUAL "sys")
+    find_dependency(PNG)
+endif()
+if (@wxUSE_LIBTIFF@ STREQUAL "sys")
+    find_dependency(TIFF)
+endif()
+
+list(APPEND TARGET_FILES wxregex wxscintilla wxbase wxcore wxgl wxxml wxhtml wxxrc wxstc wxwebview wxnet wxpropgrid wxqa wxribbon wxrichtext wxmedia wxadv wxaui)
+foreach (TARGET_FILE ${TARGET_FILES})
+    include("${CMAKE_CURRENT_LIST_DIR}/unofficial-${TARGET_FILE}Targets.cmake")
```
Comment (JackBoosY) [inline wxWidgets-config.cmake.in:36]:
```suggestion
    include("${CMAKE_CURRENT_LIST_DIR}/${TARGET_FILE}Targets.cmake")
```

**Comment by JackBoosY** [comment]
>A possible problem I noticed is that wxWidgets-config.cmake.in contains unofficial-${TARGET_FILE}Targets.cmake but the export files are not named like this. So I suspect this needs to be fixed.

That's typo, I forgot to remove this. Fixed.

> Also do the export files have to be in installdir/lib/platform/cmake or can/should it be directly in installdir/cmake?

The default path should be _installdir/cmake/LIB_NAME_ because almost every library uses this path. See [cmake doc](https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html).

**Comment by JackBoosY** [comment]
@MaartenBent Since I've merge this branch to master, some suggestions you made were not applied in this branch.
So please review again.

Thanks.

**Review by MaartenBent — COMMENTED** [review]
Thanks for applying the changes. I marked the few ones you missed.

The CMake docs suggest that the config files should be in `lib/cmake/<PackageName>`. So instead of `lib/${wxPLATFORM_LIB_DIR}/cmake` this should be `lib/cmake/wxwidgets`.

There are also some targets that are currently not included in `wxWidgets-config.cmake.in`, these are: wxexpat wxjpeg wxpng wxtiff wxzlib

Note that targets can also be turned off, and then no config file will be exported. So I think you should check if the file exists before trying to import it.

There are also some problems with the used variables in `wxWidgets-config.cmake.in`. `wxUSE_THREADS` and `wxUSE_LIBLZMA` will only get a value of `ON` or `OFF`, not `sys`. 

Also, the content of the generated file will look something like: 
```cmake
if (ON AND OFF)
    find_dependency(CURL)
endif()
```
I think it is better to add defines for all the used variables, so above code will look like:
```cmake
set(wxUSE_WEBREQUEST ON)
set(wxUSE_WEBREQUEST_CURL OFF)
if (wxUSE_WEBREQUEST AND wxUSE_WEBREQUEST_CURL)
    find_dependency(CURL)
endif()
```

And last, `wxUSE_LIBICONV` is not defined on Windows/MSVC, so the generated code has faulty syntax. This can also be fixed with the above suggestion of using defines: 
```cmake
if (UNIX AND )
    find_dependency(ICONV)
endif()
```   

Thanks again for your work in this.

**Comment by JackBoosY** [comment]
I authorize any contributor or member to modify this PR because I don’t have much time recently.

**Comment by talregev** [comment]
@vadz @MaartenBent  @TcT2k
Can you help to finish this PR?
@JackBoosY Already state it doesn't have time recently.
If you finish this PR, then it will finish the PR on vcpkg, then the opensource project I contribute can use the latest wxWidgets on our manager on linux.

**Comment by MaartenBent** [comment]
Yes, I can implement my suggestions.

But again, I have no idea if the generated config files are correct and/or useable. So I think I'll add a build option to enable generating the config files.

**Comment by talregev** [comment]
> Yes, I can implement my suggestions.
> 
> But again, I have no idea if the generated config files are correct and/or useable. So I think I'll add a build option to enable generating the config files.

Thank you! let's start from there.

**Comment by vadz** [comment]
> But again, I have no idea if the generated config files are correct and/or useable. So I think I'll add a build option to enable generating the config files.

Should we modify `samples/minimal/CMakeLists.txt` to use them? Or add another file using them? It seems like we ought to test them in the CI CMake workflow now that we have it.

**Comment by MaartenBent** [comment]
It might use them automatically when detected, or we can force it by using `find_package(wxWidgets ... CONFIG)`.

**Comment by JackBoosY** [comment]
Using `CONFIG` mode in `find_package` is a better way, because the content is the same as the build process.

**Comment by talregev** [comment]
@MaartenBent any new? Are you going to implement this PR? 
Thank you.

**Comment by MaartenBent** [comment]
Still looking into it. I can't promise when or if I'll be able to have a working version.

It seems to requires some more severe changes, e.g. using `configure_package_config_file` and `write_basic_package_version_file` instead of using `configure_file`. And then the config file probably has to define the variables that are also defined by `FindwxWidgets`.

**Comment by JackBoosY** [comment]
@MaartenBent You may need to define macros provided by `FindWxwidgets.cmake` in `config.cmake` and set them to corresponding values. Example:
In wxWidgets-targets.cmake:
```cmake
...
add_library(wxWidgets INTERFACE IMPORTED)
...
```
In wxWidgets-config.cmake:
```cmake
...
include ("${CMAKE_CURRENT_LIST_DIR}/wxWidgets-targets.cmake")
...
set(wxWidgets_FOUND 1)
get_filename_components(wxWidgets_INCLUDE_DIRS wxWidgets INTERFACE_INCLUDE_DIRECTORES)
set(wxWidgets_LIBRARIES wxWidgets)
...
```

**Comment by talregev** [comment]
Any news?

**Comment by vadz** [comment]
Sorry, no, this is unlikely to go anywhere without help from some (outside) CMake experts.

**Comment by JackBoosY** [comment]
The point is that `cotire` is seriously outdated. If we continue to use it, the automation provided by cmake will not be possible.
And if `cotire` is completely deprecated, there will be a lot of code changes.

**Comment by MaartenBent** [comment]
If `cotire` causes problems, it is possible to disable it with `wxBUILD_PRECOMP=OFF`

A long-term solution is probably trying to use [target_precompile_headers](https://cmake.org/cmake/help/latest/command/target_precompile_headers.html), when CMake is recent enough (3.16).
And I think this is more attainable than adding support for config files, since I can probably do it myself. For config files we need someone experienced with creating and/or using them.

Concerning this PR, are there parts you still like to see integrated?
The `BUILD_INTERFACE` and `INSTALL_INTERFACE` generator expressions can probably be cherry-picked, as well as the `install(EXPORT ...)` changes.
It's just the changes in `install.cmake` that I think should be based on [CMakePackageConfigHelpers](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html) instead of using `configure_file`.

**Comment by talregev** [comment]
After https://github.com/wxWidgets/wxWidgets/pull/22295 merge. We need this changes on this PR?
@JackBoosY @MaartenBent ?

**Comment by JackBoosY** [comment]
> After #22295 merge. We need these changes on this PR? @JackBoosY @MaartenBent ?

No, this PR is not needed after #22295 merged.

**Comment by talregev** [comment]
Thank you!.

**Pull request closed without merging**